examples/diff.c

```diff
@@ -86,49 +94,81 @@ int lg2_diff(git_repository *repo, int argc, char *argv[])
 	 *  * &lt;sha1&gt;
 	 *  * --cached
 	 *  * --nocache (don't use index data in diff at all)
+	 *  * --no-index &lt;file1&gt; &lt;file2&gt;
 	 *  * nothing
 	 *
 	 * Currently ranged arguments like &lt;sha1&gt;..&lt;sha2&gt; and &lt;sha1&gt;...&lt;sha2&gt;
 	 * are not supported in this example
 	 */
 
-	if (o.treeish1)
-		treeish_to_tree(&t1, repo, o.treeish1);
-	if (o.treeish2)
-		treeish_to_tree(&t2, repo, o.treeish2);
+	if (o.no_index >= 0) {
+		if (!o.treeish1 || !o.treeish2) {
+			usage("two files should be provided as arguments", NULL);
+		}
+		file1_str = read_file(o.treeish1);
+		if (file1_str == NULL) {
+			usage("file cannot be read", o.treeish1);
+		}
+		file2_str = read_file(o.treeish2);
+		if (file2_str == NULL) {
+			usage("file cannot be read", o.treeish2);
+		}
+	}
 
-	if (t1 && t2)
+	if (file1_str && file2_str) {
```

Comment:
Nitpick: I'm thinking that adding

```c
static int build_diff_from_buffers(git_diff **out, const char *b1, size_t b1len, const char *b2 size_t b2len, struct opts *diffopts);
static int build_diff_from_trees(git_diff **out, git_repository *repo, git_tree *t1, git_tree *t2, struct opts *diffopts);
```
would improve readability. That "main" function is getting pretty long, and the error path gets twisty.

Reply:
This will imply to remove the check_lg2 and check exit status for `build_diff_from_trees` so i'm not sure how to do that in a proper way